Add robot interface

hejia-zhang opened the pull request:
Add a abstract robot interface so that people can use and add different robot more easily.

hejia-zhang committed:
add robot interface

hejia-zhang committed:
add documentation for robot abstraction interface

ryanjulian commented on contrib/ros/robots/robot.py:
```diff
@@ -0,0 +1,33 @@
+        """
+        raise NotImplementedError
+
+    def set_command(self, commands):
```
Does this set a command or send a command? What is the type of the object command?

ryanjulian replied:
Shouldn't `command` be in terms of the action space?

ryanjulian commented on contrib/ros/robots/robot.py:
```diff
@@ -0,0 +1,33 @@
+
+
+class Robot(object):
+    def move_to_start_position(self):
```
Do all robots have a start position? Is this better called `reset()`? What if I want different start positions for different environments?

I'm not sure this is part of a general robot interface unless it is `reset()`

ryanjulian commented on contrib/ros/robots/robot.py:
```diff
@@ -0,0 +1,33 @@
+        """
+        raise NotImplementedError
+
+    def get_obs(self):
```
How is there `get_obs()` but no `observation_space()`?

ryanjulian replied:
Please don't abbreviate function names in public interfaces. This should be `get_observation()`.

Your API documentation should also specify the time properties of `get_observation()`. Does it return the most recent observation? Or maybe it instead returns all observations in sequence rather than the most recent one? Or maybe it guarantees that the return value of `get_observation()` is some fixed time delay after the last `send_command()`?

ryanjulian commented on contrib/ros/robots/sawyer.py:
```diff
@@ -104,15 +104,17 @@ def action_space(self):
         # For our use
         # joints[3:10]
         # set the action space depending on different control modes
+        joint_limits = rospy.wait_for_message('/robot/joint_limits',
```
Why would you retrieve the joint limits on every call to this function (blocking on a ros message), rather than getting them once and caching them?

ryanjulian commented:
Please don't submit until my comments are addressed.

hejia-zhang committed:
address ryan's comments

ryanjulian commented on contrib/ros/robots/sawyer.py:
```diff
@@ -66,7 +71,11 @@ def get_obs(self):
-    def set_command(self, commands):
+    @property
+    def observation_space(self):
+        return Box(-np.inf, np.inf, shape=self.get_observation().shape)
```
is this actually the observation space? i thought there are joint limits?

ryanjulian reviewed (commented):
is the observation space for Sawyer actually +inf/-inf?

hejia-zhang committed:
modify get_observation

hejia-zhang committed:
Add abstract robot interface

Add an abstrct robot interface so that people can use and add different robot more easily and unified.
